primitives/src/identifier.rs

```diff
@@ -0,0 +1,220 @@
+    /// Validate `bytes` is a valid CUSIP identifier, returns an instance of `Identifier` if successful
+    pub fn cusip(bytes: [u8; 9]) -> Option<AssetIdentifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(AssetIdentifier::CUSIP(bytes))
+    }
+
+    /// Validate `bytes` is a valid CINS identifier, returns an instance of `Identifier` if successful
+    pub fn cins(bytes: [u8; 9]) -> Option<AssetIdentifier> {
+        Self::cusip(bytes).map(|_| AssetIdentifier::CINS(bytes))
+    }
+
+    /// Validate `bytes` is a valid ISIN identifier, returns an instance of `Identifier` if successful
+    pub fn isin(bytes: [u8; 12]) -> Option<AssetIdentifier> {
+        let (s1, s2) = bytes
+            .iter()
+            .copied()
+            .map(byte_value)
+            .flat_map(|b| if b > 9 { vec![b / 10, b % 10] } else { vec![b] })
+            .rev()
+            .enumerate()
+            .fold((0, 0), |(mut s1, mut s2), (i, digit)| {
+                if i % 2 == 0 {
+                    s1 += digit;
+                } else {
+                    s2 += 2 * digit;
+                    if digit >= 5 {
+                        s2 -= 9;
+                    }
+                }
+                (s1, s2)
+            });
+        ((s1 + s2) % 10 == 0).then_some(AssetIdentifier::ISIN(bytes))
+    }
+
+    /// Validate `bytes` is a valid LEI identifier, returns an instance of `Identifier` if successful
+    pub fn lei(bytes: [u8; 20]) -> Option<AssetIdentifier> {
+        bytes[..18]
+            .try_into()
+            .ok()
+            .map(lei_checksum)
+            .filter(|hash| *hash == (bytes[18] - b'0') * 10 + (bytes[19] - b'0'))
+            .map(|_| AssetIdentifier::LEI(bytes))
+    }
+
+    /// Ensures the identifier is valid.
+    /// Mainly used for validating manual constructions of the enum (user input).
+    pub fn validate(self) -> Option<Self> {
+        match self {
+            AssetIdentifier::CUSIP(bytes) => Self::cusip(bytes),
+            AssetIdentifier::CINS(bytes) => Self::cins(bytes),
+            AssetIdentifier::ISIN(bytes) => Self::isin(bytes),
+            AssetIdentifier::LEI(bytes) => Self::lei(bytes),
+            AssetIdentifier::EMPTY => None,
+        }
+    }
+}
+
+fn cusip_checksum(bytes: &[u8]) -> u8 {
+    let total: usize = bytes
+        .iter()
+        .copied()
+        .map(byte_value)
+        .enumerate()
+        .map(|(i, v)| v << (i % 2))
+        .map(|v| (v / 10) + v % 10)
+        .map(|x| x as usize)
+        .sum();
+    ((10 - (total % 10)) % 10) as u8
+}
+
+fn lei_checksum(bytes: [u8; 18]) -> u8 {
+    let mut i = 0;
+    let total = bytes
+        .iter()
+        .copied()
+        .rev()
+        .map(byte_value)
+        .map(|x| x as u128)
+        .fold(0, |mut total, b| {
+            total += b * 10u128.pow(i as u32);
+            if b > 9 {
+                i += 2;
+            } else {
+                i += 1;
+            }
+            total
+        });
```

Comment:
A final bit of refactoring to something more functional:

```suggestion
        .fold(0, |mut total, b| {
            total += b * 10u128.pow(i as u32);
            i += if b > 9 { 2 } else { 1 };
            total
        });
```